clang/lib/CConv/ConstraintResolver.cpp

```diff
@@ -214,7 +214,10 @@ CVarSet
         return PVConstraintFromType(TypE);
       }
       // NULL
-    } else if (isNULLExpression(E, *Context)) {
+      // Special handling for casts of null is required to enable rewriting
+      // statements such as int *x = (int*) 0. If this was handled as a
+      // normal null expression, the cast would never be visited.
+    } else if (!isa<ExplicitCastExpr>(E) && isNULLExpression(E, *Context)) {
```

Comment:
I was wondering about NULL expressions. Did you confirm that this change works on vsftpd and other code that uses standard headers?

Reply:
I don't understand what exactly you're asking about here. The PR can convert vstpd and the output compiles. Any specific example you're wondering about?

Reply:
I was wondering about expressions like `(int **)0` or `(int *)0` or `(char ***)0`. All of these effectively mean `'NULL` at particular types. Your check here is saying "if this is not an explicit cast but is a null expression return the empty set." But NULL expressions typically *are* an explicit cast, i.e., `(void *)0`. They are not usually just `0` with no cast after it. So that makes me wonder whether this case will now ever actually be reached.

Reply:
This case isn't reached if there are casts on the null expression, but `isNULLExpression` still returns true for  `0`, so it gets hit on those.